source/slang/slang-ir.cpp

```diff
@@ -4325,11 +4327,114 @@ namespace Slang
+                else if(auto rayPayloadIn = i->findDecoration<IRVulkanRayPayloadInDecoration>())
+                {
+                    storeLocationToRayPayloadVariable(int(getIntVal(rayPayloadIn->getOperand(0))), i);
+                }
+                else if(auto hitObject = i->findDecoration<IRVulkanHitObjectAttributesDecoration>())
+                {
+                    storeLocationToRayAttributeVariable(int(getIntVal(hitObject->getOperand(0))), i);
+                }
+                else if(auto callable = i->findDecoration<IRVulkanCallablePayloadDecoration>())
+                {
+                    storeLocationToRayCallableVariable(int(getIntVal(callable->getOperand(0))), i);
+                }
+                else if(auto callableIn = i->findDecoration<IRVulkanCallablePayloadInDecoration>())
+                {
+                    storeLocationToRayCallableVariable(int(getIntVal(callableIn->getOperand(0))), i);
+                }
+            }
+
+        }
+    }
+
+    IRInst* IRModule::getRayVariableFromLocation(IRInst* payloadVariable, Slang::IROp op, DiagnosticSink* sink)
+    {
+        IRBuilder builder(payloadVariable);
+        IRInst** varLayoutPointsTo = nullptr;
+        int intLitValue = -1;
+        IRIntLit* intLit = as<IRIntLit>(payloadVariable);
+        if(intLit){
+            intLitValue = int(intLit->getValue());
+            if (kIROp_SPIRVAsmOperandRayPayloadFromLocation == op)
+            {
+                varLayoutPointsTo = this->getRayPayloadVariableFromLocation(intLitValue);
+            }
+            else if (kIROp_SPIRVAsmOperandRayAttributeFromLocation == op)
+            {
+                varLayoutPointsTo = this->getRayAttributeVariableFromLocation(intLitValue);
+            }
+            else if (kIROp_SPIRVAsmOperandRayCallableFromLocation == op)
+            {
+                SLANG_ASSERT(kIROp_SPIRVAsmOperandRayCallableFromLocation == op); //final case
+                varLayoutPointsTo = this->getRayCallableVariableFromLocation(intLitValue);
+            }
+        }
+        else{
+            sink->diagnose(payloadVariable, Diagnostics::expectedIntegerConstantNotConstant);
+        }
+
+        IRInst* resultVariable;
+        if (!varLayoutPointsTo)
+        {
+            // if somehow the location tied variable is missing and an error was not thrown by the compiler
+            resultVariable = builder.getIntValue(builder.getIntType(), 0);
+            sink->diagnose(payloadVariable, Diagnostics::expectedRayTracingPayloadObjectAtLocationButMissing, intLitValue);
+        }
+        else {
+            resultVariable = *varLayoutPointsTo;
+        }
+        return resultVariable;
+    }
+
+    void IRModule::storeLocationToRayPayloadVariable(int location, IRInst* inst)
+    {
+        // Try to remove duplicates since IR will try to remove/re-add globals 
+        // sometimes to move them around, this case needs to be handled
+        m_RayLocationToPayloads.set(location, inst);
```

Comment:
We shouldn't add a new field to `IRModule` since we don't need this for more than the span of a single pass.
Instead we should be able to reconstruct the dictionary from all the decorations.